feat(TFTrainer): improve logging

borisdayma opened the pull request:
This PR supersedes PR #4756 
I wanted initially to refactor logging between `Trainer` and `TFTrainer` but there seems to be too many differences between them for it to make sense (distributed training, logging methods, `wandb.watch`…).

Logging is now handled directly within each Trainer and all comments from previous PR have been applied here.

Notes:

* `TFTrainer`: @jplu There should be the same modifications we discussed. I just moved the check of `global_step` within logging for better readability.
* `TFTrainer._log` does not have a tqdm iterator argument (unlike in `Trainer`) since you're not using it at the moment but it could be useful in the future
* `Trainer`: main update is to handle the case where we do only evaluations (no training)

borisdayma commented:
Integration test error does not seem to be related to this PR.

@jplu On another note, I got strange results while using `run_tf_glue` as learning rate goes quickly to 0.

Command:
`run_tf_glue.py --model_name_or_path bert-base-cased --task_name MRPC --do_train --do_eval --max_seq_length 128 --per_device_train_batch_size 32 --learning_rate 2e-5 --num_train_epochs 3.0 --output_dir /tmp/MRPC/ --overwrite_output_dir --logging_dir log --evaluate_during_training --eval_steps 50 --logging_steps 10`

Graph
![image](https://user-images.githubusercontent.com/715491/84448088-5e9c2600-ac0f-11ea-9994-bdb64c84c483.png)

W&B Run: https://app.wandb.ai/borisd13/huggingface/runs/21rxop7c

jplu commented:
Good you have restarted a new PR!

Humm for the loss that drop quickly to 0 I think it might come from your side because I get normal evolution even after multiple runs. Here an output:

```
06/12/2020 10:05:38 - INFO - transformers.trainer_tf -   ***** Running training *****
06/12/2020 10:05:38 - INFO - transformers.trainer_tf -     Num examples = 3668
06/12/2020 10:05:38 - INFO - transformers.trainer_tf -     Num Epochs = 3
06/12/2020 10:05:38 - INFO - transformers.trainer_tf -     Total optimization steps = 29
WARNING:tensorflow:From /home/jplu/transformers/src/transformers/trainer_tf.py:355: StrategyBase.experimental_run_v2 (from tensorflow.python.distribute.distribute_lib) is deprecated and will be removed in a future version.
Instructions for updating:
renamed to `run`
06/12/2020 10:05:38 - WARNING - tensorflow -   From /home/jplu/transformers/src/transformers/trainer_tf.py:355: StrategyBase.experimental_run_v2 (from tensorflow.python.distribute.distribute_lib) is deprecated and will be removed in a future version.
Instructions for updating:
renamed to `run`
WARNING:tensorflow:From /opt/anaconda3/envs/jplu-transformers/lib/python3.7/site-packages/tensorflow/python/ops/resource_variable_ops.py:1817: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.
Instructions for updating:
If using Keras pass *_constraint arguments to layers.
06/12/2020 10:05:47 - WARNING - tensorflow -   From /opt/anaconda3/envs/jplu-transformers/lib/python3.7/site-packages/tensorflow/python/ops/resource_variable_ops.py:1817: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.
Instructions for updating:
If using Keras pass *_constraint arguments to layers.
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
06/12/2020 10:06:17 - INFO - tensorflow -   Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
06/12/2020 10:06:17 - INFO - tensorflow -   Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
06/12/2020 10:06:53 - INFO - tensorflow -   Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
06/12/2020 10:06:53 - INFO - tensorflow -   Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:batch_all_reduce: 201 all-reduces with algorithm = nccl, num_packs = 1
06/12/2020 10:08:32 - INFO - tensorflow -   batch_all_reduce: 201 all-reduces with algorithm = nccl, num_packs = 1
INFO:tensorflow:batch_all_reduce: 201 all-reduces with algorithm = nccl, num_packs = 1
06/12/2020 10:08:48 - INFO - tensorflow -   batch_all_reduce: 201 all-reduces with algorithm = nccl, num_packs = 1
06/12/2020 10:09:46 - INFO - transformers.trainer_tf -   Epoch 1 Step 10 Train Loss 0.6610
06/12/2020 10:10:02 - INFO - transformers.trainer_tf -   Epoch 1 Step 20 Train Loss 0.5626
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
06/12/2020 10:10:46 - INFO - tensorflow -   Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
06/12/2020 10:10:46 - INFO - tensorflow -   Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
06/12/2020 10:11:15 - INFO - transformers.trainer_tf -   Epoch 2 Step 30 Train Loss 0.5585
06/12/2020 10:11:31 - INFO - transformers.trainer_tf -   Epoch 2 Step 40 Train Loss 0.5492
06/12/2020 10:11:47 - INFO - transformers.trainer_tf -   ***** Running Evaluation *****
06/12/2020 10:11:47 - INFO - transformers.trainer_tf -     Batch size = 32
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
06/12/2020 10:11:53 - INFO - tensorflow -   Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
06/12/2020 10:11:53 - INFO - tensorflow -   Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
06/12/2020 10:12:04 - INFO - tensorflow -   Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
06/12/2020 10:12:04 - INFO - tensorflow -   Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
06/12/2020 10:12:08 - INFO - transformers.trainer_tf -   Epoch 2 Step 50 Validation Metrics {'eval_eval_loss': 0.5300102, 'eval_eval_acc': 0.7230392156862745, 'eval_eval_f1': 0.8274809160305343, 'eval_eval_acc_and_f1': 0.7752600658584043, 'learning_rate': 0.0}
06/12/2020 10:12:08 - INFO - transformers.trainer_tf -   Epoch 2 Step 50 Train Loss 0.6273
06/12/2020 10:13:21 - INFO - transformers.trainer_tf -   Epoch 3 Step 60 Train Loss 0.6168
06/12/2020 10:13:37 - INFO - transformers.trainer_tf -   Epoch 3 Step 70 Train Loss 0.5616
06/12/2020 10:13:53 - INFO - transformers.trainer_tf -   Epoch 3 Step 80 Train Loss 0.5551
06/12/2020 10:14:05 - INFO - transformers.trainer_tf -   Saving model in /tmp/MRPC/
```

I used the exact same command line than yours.

borisdayma commented:
I seem to be having a different value of `self.train_steps`.

```
Num examples = 3668
Num Epochs = 3
Total optimization steps = 115
```
Here is my full log output: https://app.wandb.ai/borisd13/huggingface/runs/1u5spvau/logs

Actually I just checked and I'm getting the exact same problem in the `master` branch so i guess the issue I'm having is independent from this PR.

Let me know if you need any modification on this PR.

jplu commented:
Sorry I don't see what is the issue in your logs, all the values seems ok. I get 29 steps because I run over 4 gpus.

borisdayma commented:
Here is a gist with full output: https://gist.github.com/borisdayma/62fd9338aae4f373a9c0709a8961f5bc

It's probably independent. I could file a separate issue.

jplu commented:
I still don't see what is the issue sorry, everything looks normal.

borisdayma commented:
@jplu actually I noticed you have the same issue.

```
06/12/2020 10:12:08 - INFO - transformers.trainer_tf -   Epoch 2 Step 50 Validation Metrics {'eval_eval_loss': 0.5300102, 'eval_eval_acc': 0.7230392156862745, 'eval_eval_f1': 0.8274809160305343, 'eval_eval_acc_and_f1': 0.7752600658584043, 'learning_rate': 0.0}
```

In current master, learning_rate is displayed only at eval. It should not be 0 at this specific step.
If you confirm let me know if you want me to file an issue.

In any case, since it's independent from this PR, let me know if you approve it.
It's always difficult to stay in sync with master so I'd like to make all necessary changes as soon as possible since it's a difficult PR.

jplu commented:
This is normal, it is the decay of the LR so at some point it gets 0. It is ok. I will take some time to review your PR this weekend.

jplu commented:
Oh I see what you mean, yes it gets to 0 at the end of the first epoch, and it shouldn't, it is fixed from my side, PR will be here so no worries we can focus on your logging improvement. Thanks!!

jplu commented on src/transformers/trainer_tf.py:
```diff
@@ -208,6 +248,22 @@ def _prediction_loop(
         return PredictionOutput(predictions=preds, label_ids=label_ids, metrics=metrics)
 
+    def _log(self, logs: Dict[str, float]) -> None:
+        if self.epoch_logging is not None:
+            logs["epoch"] = self.epoch_logging
+        if self.global_step is None:
+            # when logging evaluation metrics without training
+            self.global_step = 0
```
Here you are also logging `global_step` and `epoch_logging` during the evaluation, it should not. Then these 5 lines can be removed.

borisdayma replied:
Because we use `eval_steps` arg, which may be different than `logging_steps`, we need to make sure we log the epoch as it could be associated to a different step.

My example gives below graph:

* where we have less evaluation steps (that could also be different from the logging steps).
* where we also log final evaluation metrics at the end of training -> you can see `eval_loss` is logged further than `loss`

![image](https://user-images.githubusercontent.com/715491/84574274-fd896500-ad6a-11ea-9a96-06e81f211c9d.png)

The step is barely needed as default x-axis and to make sure we have sequential logs (as with Tensorboard) and we just use the current training step for convenience.

When doing evaluation, we have access to `global_step` and `epoch_logging` as they are being saved during the training directly in the object.

I can set `global_step` in `__init__` so it's never `None` and then I can remove lines 254-256.

I still need to log `epoch` in eval (to make the graph above).
In the current logic, I don't log it at all in the case that I'm using `TFTrainer` only for evaluation (no training) as I thought it would not be useful but if you want I can init it to 0 as well. I would still need to set `logs["epoch"] though.

jplu replied:
Sorry, what I meant is that `logs["epoch"]` can be set in the `train` method and `global_step` in `__init__`. As if you run only the trainer in eval/predict mode you don't need to have a graph depending of epochs and if you set `global_step` in `__init__` you don't have to do this `if`.

borisdayma commented:
@jplu I addressed your comments. Let me know if I understood correctly:

* `global_step` moved to `__init__`
* `epoch` added to logs directly within training loop
* In addition, I also added `epoch` in eval loop when called from training loop. If we log training loss every 20 steps and we log evaluation metrics every 50 steps, we need to make sure we add `epoch`

I'm going to add a comment for a further simplification if you'd like.

borisdayma commented on src/transformers/trainer_tf.py:
```diff
@@ -269,44 +327,40 @@ def train(self) -> None:
-                if self.args.debug:
-                    with self.tb_writer.as_default():
-                        tf.summary.scalar("loss", training_loss, step=step)
+        self.epoch_logging = 0
```
This could also be added to `__init__` along with `self.global_step`.
Note that if that's the case, it means we will log an epoch of `0` when doing evaluation with no training (see previous comment).

jplu replied:
Yes, indeed. If you removes the `_log` call in the eval method there is no more issue.

jplu replied:
Exactly, but without to make it appear in the logs when running eval/predict mode.

borisdayma replied:
We want `evaluate` to always log metrics, wherever it's called from (training loop or outside in evaluation only from example scripts).

If `self.global_step` in in `__init__`, it is harder to differentiate whether it's being called from training loop or outside of it.

We could check if `self.epoch_logging == 0` but it would not be correct because we could want to do the following:
* initialize `Trainer` with a pre-trained model
* call `evaluate` to log evaluation metrics for reference (at this point `self.epoch_logging = 0`)
* fine-tune through training loop (where `epoch_logging` will slowly increase)
* call `evaluate` to log final evaluation metrics

In order to see my evaluation metrics on a same graph with `epoch` as x-axis, I would need to have it logged at each step.
It seems a more universal solution to just log `epoch` all the time but let me know if you have any other idea.

jplu replied:
The role of the `_logs` method is just to log the content of `logs` dict and must be independent of how it is used. Step and epoch should not be taken into account, they both have a value during the training and they both equals 0 during evaluation.

The trainer has to run an evaluation at each `% eval_step` and output logs at each `% logging_step`. If someone wants to have an evaluation at the end of the training, add the `--do_eval` parameter will do the job.

borisdayma commented on src/transformers/trainer_tf.py:
```diff
@@ -218,6 +270,12 @@ def evaluate(
 
+        logs = {**output.metrics}
+        # add epoch when running evaluate from training loop
+        if self.epoch_logging is not None:
```
Here I made a mistake and it should be `if hasattr(self, "epoch_logging")`.

An easier solution (if we move `epoch_logging` to `__init__`) would be to avoid this verification at all since this attribute will always exist (and there's no harm in using an epoch of `0` when doing no training).

Let me know your preference.

jplu replied:
You should not log anything here. That's why there is the logging_step and eval_step parameters. Then the evaluate method should stay as it was. And log only in the train metrhod.

jplu replied:
No epoch should not appear in the logs even when you do only eval.

borisdayma replied:
See my next comment. It is hard to identify when I'm in `eval/predict` only and there could be some issues when wanting to eval prior training.

jplu commented on src/transformers/trainer.py:
```diff
@@ -556,6 +555,9 @@ def train(self, model_path: Optional[str] = None):
     def _log(self, logs: Dict[str, float], iterator: Optional[tqdm] = None) -> None:
         if self.epoch is not None:
             logs["epoch"] = self.epoch
+        if self.global_step is None:
+            # when logging evaluation metrics without training
+            self.global_step = 0
         if self.tb_writer:
```
Same thing here.

borisdayma replied:
Here I'm pinging @julien-c because I initially had it in `__init__` and he preferred to move it within the training loop.
I've been trying to apply the same logic with `TFTrainer` but I have no issue in doing something different.

jplu replied:
We don't expect to have both trainer exactly the same. The logging part hasn't been a lot discussed yet, it is currently the main difference and I think we should take some time to properly rethink that part.

jplu commented on src/transformers/trainer_tf.py:
```diff
@@ -269,44 +327,40 @@ def train(self) -> None:
+                self.epoch_logging = epoch_iter - 1 + (step + 1) / self.train_steps
 
-                if self.args.evaluate_during_training and step % self.args.eval_steps == 0:
+                if self.args.debug:
```
This is wrong. This if should be run only when we want to evaluate during training and when the step is a modulo of the `eval_steps`.

jplu commented on src/transformers/trainer_tf.py:
```diff
@@ -269,44 +327,40 @@ def train(self) -> None:
-                        tf.summary.scalar("loss", training_loss, step=step)
+        self.epoch_logging = 0
 
-                if step == 1 and self.args.debug:
-                    with self.tb_writer.as_default():
-                        tf.summary.trace_export(name="training", step=step, profiler_outdir=self.args.logging_dir)
```
Why remove this?

jplu commented on src/transformers/trainer_tf.py:
```diff
@@ -269,44 +327,40 @@ def train(self) -> None:
-
-                    logger.info("Epoch {} Step {} Validation Metrics {}".format(epoch, step, logs))
+                    logs["loss"] = training_loss.numpy()
+                    logs["epoch"] = self.epoch_logging
+                    self._log(logs)
 
+                if self.global_step == 1 and self.args.debug:
                     with self.tb_writer.as_default():
-                        for k, v in logs.items():
-                            tf.summary.scalar(k, v, step=step)
+                        tf.summary.trace_export(
+                            name="training", step=self.global_step, profiler_outdir=self.args.logging_dir
+                        )
 
-                    self.tb_writer.flush()
+                if self.args.evaluate_during_training and self.global_step % self.args.eval_steps == 0:
+                    self.evaluate()
 
-                if step % self.args.logging_steps == 0:
-                    logger.info("Epoch {} Step {} Train Loss {:.4f}".format(epoch, step, training_loss.numpy()))
+                if self.global_step % self.args.logging_steps == 0:
+                    logs = {}
+                    logs["loss"] = training_loss.numpy()
+                    logs["learning_rate"] = lr_scheduler(self.global_step).numpy()
+                    logs["epoch"] = self.epoch_logging
+                    self._log(logs)
 
-                if step % self.args.save_steps == 0:
+                if self.global_step % self.args.save_steps == 0:
                     ckpt_save_path = self.model.ckpt_manager.save()
-                    logger.info("Saving checkpoint for step {} at {}".format(step, ckpt_save_path))
+                    logger.info("Saving checkpoint for step {} at {}".format(self.global_step, ckpt_save_path))
 
-                if step % self.train_steps == 0:
+                if self.global_step % self.train_steps == 0:
```
Please keep the `if`s in the same order than it was before.

borisdayma replied:
Not sure I understand this comment...
I didn't change any order?

jplu replied:
In the diff view I see the order changed, but not in the file so it is ok, the diff view can be error prone sometime^^

jplu commented on src/transformers/trainer_tf.py:
```diff
@@ -269,44 +327,40 @@ def train(self) -> None:
 
-                    self.tb_writer.flush()
+                if self.args.evaluate_during_training and self.global_step % self.args.eval_steps == 0:
+                    self.evaluate()
```
You should keep the return of evaluate and log it. And not in the evaluate method.

jplu commented:
Ok, we almost see the end of the tunnel. Just few comments yet :)

borisdayma commented:
The way it is now, `epoch` and `global_step` will just be 0 in eval mode only.

julien-c commented on src/transformers/trainer_tf.py:
```diff
@@ -14,6 +14,23 @@
+
+
+def is_wandb_available():
+    return _has_wandb
```
I'm ok with moving this block to trainer_utils (it can be in a following PR)

borisdayma replied:
Sounds good, I'll propose it in a follow-up PR because we still need to import wandb in the other modules so I need to think of a clean approach.
